zokrates/numbertheory.py

```diff
@@ -0,0 +1,849 @@
+
+    for p in smallprimes:
+        print("Testing square_root_mod_prime for modulus p = %d." % p)
+        squares = []
+
+        for root in range(0, 1 + p // 2):
+            sq = (root * root) % p
+            squares.append(sq)
+            calculated = square_root_mod_prime(sq, p)
+            if (calculated * calculated) % p != sq:
+                error_tally = error_tally + 1
+                print(
+                    "Failed to find %d as sqrt( %d ) mod %d. Said %d."
+                    % (root, sq, p, calculated)
+                )
+
+        for nonsquare in range(0, p):
+            if nonsquare not in squares:
+                try:
+                    calculated = square_root_mod_prime(nonsquare, p)
+                except SquareRootError:
+                    pass
+                else:
+                    error_tally = error_tally + 1
+                    print(
+                        "Failed to report no root for sqrt( %d ) mod %d."
+                        % (nonsquare, p)
+                    )
+
+    # Test the jacobi function:
+    for m in range(3, 400, 2):
+        print("Testing jacobi for modulus m = %d." % m)
+        if is_prime(m):
+            squares = []
+            for root in range(1, m):
+                if jacobi(root * root, m) != 1:
+                    error_tally = error_tally + 1
+                    print("jacobi( %d * %d, %d ) != 1" % (root, root, m))
+                squares.append(root * root % m)
+            for i in range(1, m):
+                if not i in squares:
+                    if jacobi(i, m) != -1:
+                        error_tally = error_tally + 1
+                        print("jacobi( %d, %d ) != -1" % (i, m))
+        else:  # m is not prime.
+            f = factorization(m)
+            for a in range(1, m):
+                c = 1
+                for i in f:
+                    c = c * jacobi(a, i[0]) ** i[1]
+                if c != jacobi(a, m):
+                    error_tally = error_tally + 1
+                    print("%d != jacobi( %d, %d )" % (c, a, m))
+
+    # Test the inverse_mod function:
+    print("Testing inverse_mod . . .")
+    import random
+
+    n_tests = 0
+    for i in range(100):
+        m = random.randint(20, 10000)
+        for j in range(100):
+            a = random.randint(1, m - 1)
+            if gcd(a, m) == 1:
+                n_tests = n_tests + 1
+                inv = inverse_mod(a, m)
+                if inv <= 0 or inv >= m or (a * inv) % m != 1:
+                    error_tally = error_tally + 1
+                    print("%d = inverse_mod( %d, %d ) is wrong." % (inv, a, m))
+    assert n_tests > 1000
+    print(n_tests, " tests of inverse_mod completed.")
+
+    class FailedTest(Exception):
+        pass
+
+    print(error_tally, "errors detected.")
+    if error_tally != 0:
+        raise FailedTest("%d errors detected" % error_tally)
+
+
+if __name__ == "__main__":
+    __main__()
```

Comment:
FYI: I haven't done any review on this file.

Reply:
makes sense. added a reference to the original file (https://github.com/warner/python-ecdsa/blob/master/src/ecdsa/numbertheory.py)